File: app/src/main/java/com/spiderbiggen/randomchampionselector/ui/adapters/ChampionAdapter.java
package com.spiderbiggen.randomchampionselector.ui.adapters;

import android.content.Context;
import android.graphics.Bitmap;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.spiderbiggen.randomchampionselector.R;
import com.spiderbiggen.randomchampionselector.ddragon.DDragon;
import com.spiderbiggen.randomchampionselector.model.Champion;
import com.spiderbiggen.randomchampionselector.model.ImageType;

import java.io.IOException;
import java.util.List;

import static com.spiderbiggen.randomchampionselector.ddragon.DDragon.createDDragon;

public class ChampionAdapter extends RecyclerView.Adapter<ChampionAdapter.ViewHolder> {

    private static final String TAG = ChampionAdapter.class.getSimpleName();
    private final Context context;

    private final View.OnClickListener clickListener;
    private List<Champion> champions;

    public ChampionAdapter(final Context context, final List<Champion> champions, final View.OnClickListener clickListener) {
        this.context = context;
        this.champions = champions;
        this.clickListener = clickListener;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        // create a new view
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.list_champion_item, parent, false);
        v.setOnClickListener(clickListener);
        return new ViewHolder(v);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        Champion champion = champions.get(position);

        DDragon dDragon = createDDragon(context);
        try {
            Bitmap backGround = dDragon.getChampionBitmap(champion, ImageType.SPLASH);
            holder.backGroundView.setImageBitmap(backGround);
        } catch (IOException e) {
            Log.e(TAG, "onBindViewHolder: ", e);
        }
        holder.nameView.setText(champion.getName());
        holder.titleView.setText(champion.getCapitalizedTitle());
    }

    @Override
    public int getItemCount() {
        return champions != null ? champions.size() : 0;
    }

    public void setChampions(List<Champion> champions) {
        this.champions = champions;
        notifyDataSetChanged();
    }

    public Champion getChampion(int position) {
        if (position > getItemCount() || position < 0) return null;
        return champions.get(position);
    }

    static class ViewHolder extends RecyclerView.ViewHolder {

        private ImageView backGroundView;
        private TextView nameView;
        private TextView titleView;

        private ViewHolder(View itemView) {
            super(itemView);
            nameView = itemView.findViewById(R.id.championName);
            titleView = itemView.findViewById(R.id.championTitle);
            backGroundView = itemView.findViewById(R.id.champion_background);
        }
    }
}

File: app/src/main/java/com/spiderbiggen/randomchampionselector/views/adapters/ChampionAdapter.kt
package com.spiderbiggen.randomchampionselector.views.adapters

import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.spiderbiggen.randomchampionselector.R
import com.spiderbiggen.randomchampionselector.data.cache.BitmapCache
import com.spiderbiggen.randomchampionselector.data.onMainThread
import com.spiderbiggen.randomchampionselector.domain.Champion
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.io.IOException


class ChampionAdapter(champions: Collection<Champion>, private val clickListener: View.OnClickListener) : RecyclerView.Adapter<ChampionAdapter.ViewHolder>(){

    private val champions = champions.toMutableList()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        // create a new view
        val v = LayoutInflater.from(parent.context).inflate(R.layout.list_champion_item, parent, false)
        v.setOnClickListener(clickListener)
        return ViewHolder(v)
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, position: Int) {
        viewHolder.champion = champions[position]
    }

    override fun getItemCount(): Int = champions.size

    fun setChampions(champions: Collection<Champion>) {
        this.champions.clear()
        this.champions.addAll(champions)
        notifyDataSetChanged()
    }

    fun getChampion(position: Int): Champion? = champions.getOrNull(position)

    class ViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView), CoroutineScope by CoroutineScope(Dispatchers.Default) {
        private val imageView: ImageView = itemView.findViewById(R.id.champion_splash)

        private val nameView: TextView = itemView.findViewById(R.id.champion_name)
        private val titleView: TextView = itemView.findViewById(R.id.champion_title)

        internal var champion: Champion? = null
            set(value) {
                if (value == null) throw NullPointerException()
                updateImage(value)
                nameView.text = value.name
                titleView.text = value.capitalizedTitle
                field = value
            }

        private fun updateImage(champion: Champion) {
            launch {
                try {
                    val bitmap = BitmapCache.loadBitmap(champion)
                    onMainThread { imageView.setImageBitmap(bitmap) }
                } catch (e: IOException) {
                    onMainThread { loadImageFailure(e.message) }
                }
            }
        }

        private fun loadImageFailure(message: String?) {
            imageView.setImageBitmap(null)
            Log.e("ChampionAdapter", "error $message")
        }
    }
}
